packages/eslint-plugin/src/rules/no-unnecessary-type-constraint.ts

```diff
@@ -0,0 +1,116 @@
+import { TSESTree } from '@typescript-eslint/experimental-utils';
+import * as semver from 'semver';
+import * as ts from 'typescript';
+import * as util from '../util';
+
+type MakeRequired<Base, Key extends keyof Base> = Omit<Base, Key> &
+  Required<Pick<Base, Key>>;
+
+type TypeParameterWithConstraint = MakeRequired<
+  TSESTree.TSTypeParameter,
+  'constraint'
+>;
+
+type KeywordFilter = (type: ts.Type) => boolean;
+
+const is3dot5 = semver.satisfies(
+  ts.version,
+  `>= 3.5.0 || >= 3.5.1-rc || >= 3.5.0-beta`,
+  {
+    includePrerelease: true,
+  },
+);
+
+const is3dot9 =
+  is3dot5 &&
+  semver.satisfies(ts.version, `>= 3.9.0 || >= 3.9.1-rc || >= 3.9.0-beta`, {
+    includePrerelease: true,
+  });
+
+export default util.createRule({
+  name: 'no-unnecessary-type-constraint',
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description: 'Disallows unnecessary constraints on generic types',
+      recommended: false,
+      requiresTypeChecking: true,
+      suggestion: true,
+    },
+    fixable: 'code',
+    messages: {
+      unnecessaryConstraint:
+        'Constraining a generic type to {{constraint}} does nothing and is unnecessary.',
+    },
+    schema: [],
+    type: 'suggestion',
+  },
+  defaultOptions: [],
+  create(context) {
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    const keywordFilters: [KeywordFilter, string][] = [];
+    if (is3dot5) {
+      keywordFilters.push([util.isTypeUnknownType, 'unknown']);
+
+      if (is3dot9) {
+        keywordFilters.push([util.isTypeAnyType, 'any']);
+      }
+    }
+
+    if (!keywordFilters.length) {
+      return {};
```

Comment:
😕 is there a way I can specify different versions of TypeScript in unit tests?

Reply:
We don't yet have any test tooling around specifying TS versions (for any package).
On the to-do list but nobody has ever prioritised it.